File: packages/swt3-ai/swt3_ai/adapters/openai.py
"""SWT3 AI Witness SDK — OpenAI Adapter (ChainProxy).

Implements a recursive proxy that transparently wraps the OpenAI client
so that `client.chat.completions.create()` is intercepted for witnessing
while all other attributes pass through untouched.

The ChainProxy follows the attribute chain:
    client.chat         → ChainProxy (passthrough)
    client.chat.completions → ChainProxy (passthrough)
    client.chat.completions.create() → INTERCEPTED

On interception:
    1. Capture the prompt messages for hashing
    2. Start the latency timer
    3. Call the real create() method
    4. Extract factors from the response
    5. Build an InferenceRecord
    6. Hand to Witness.record() (non-blocking buffer)
    7. Return the original response to the developer UNTOUCHED

The developer's code sees zero difference from using the raw OpenAI client.
"""

from __future__ import annotations

import time
from typing import Any, Callable, Dict, Set, TYPE_CHECKING

from ..types import InferenceRecord
from ..fingerprint import sha256_truncated

if TYPE_CHECKING:
    from ..witness import Witness


# Methods to intercept at the terminal level
INTERCEPTED_METHODS: Dict[str, Set[str]] = {
    # path → method names to intercept
    "chat.completions": {"create"},
    # Future: "embeddings": {"create"},
    # Future: "images": {"generate"},
}

# Map attribute chains to their dotted path
# e.g., client.chat.completions → "chat.completions"


def wrap_openai(client: Any, witness: "Witness") -> "_OpenAIProxy":
    """Wrap an OpenAI client with transparent witnessing."""
    return _OpenAIProxy(client, witness, path="")


class _OpenAIProxy:
    """Recursive proxy that follows the OpenAI client attribute chain.

    At each level, __getattr__ checks if the resolved path has interceptable
    methods. If the next attribute IS an interceptable method, it returns a
    wrapped callable. Otherwise, it returns another proxy for the next level.
    """

    __slots__ = ("_target", "_witness", "_path")

    def __init__(self, target: Any, witness: "Witness", path: str) -> None:
        object.__setattr__(self, "_target", target)
        object.__setattr__(self, "_witness", witness)
        object.__setattr__(self, "_path", path)

    def __getattr__(self, name: str) -> Any:
        target = object.__getattribute__(self, "_target")
        witness = object.__getattribute__(self, "_witness")
        path = object.__getattribute__(self, "_path")

        # Resolve the real attribute from the target
        real_attr = getattr(target, name)

        # Build the dotted path for this level
        current_path = f"{path}.{name}" if path else name

        # Check if this attribute is an interceptable method
        parent_path = path  # e.g., "chat.completions"
        if parent_path in INTERCEPTED_METHODS and name in INTERCEPTED_METHODS[parent_path]:
            # This IS the method to intercept — wrap it
            return _make_interceptor(real_attr, witness, name, parent_path)

        # Check if this path could lead to interceptable methods
        # e.g., "chat" could lead to "chat.completions"
        has_children = any(p.startswith(current_path) for p in INTERCEPTED_METHODS)

        if has_children:
            # Keep proxying deeper
            return _OpenAIProxy(real_attr, witness, current_path)

        # No interception needed — return the real attribute directly
        return real_attr

    def __repr__(self) -> str:
        target = object.__getattribute__(self, "_target")
        return f"<WitnessProxy({type(target).__name__})>"


def _make_interceptor(
    real_method: Callable[..., Any],
    witness: "Witness",
    method_name: str,
    path: str,
) -> Callable[..., Any]:
    """Create an interceptor that wraps an OpenAI API method."""

    def interceptor(*args: Any, **kwargs: Any) -> Any:
        # ── Pre-call: capture prompt for hashing ──
        messages = kwargs.get("messages", args[0] if args else [])
        model = kwargs.get("model", "unknown")

        # Hash the prompt (all message contents concatenated)
        prompt_text = _extract_prompt_text(messages)
        prompt_hash = sha256_truncated(prompt_text)

        # ── Call the real method and measure latency ──
        start = time.monotonic()
        response = real_method(*args, **kwargs)
        elapsed_ms = int((time.monotonic() - start) * 1000)

        # ── Post-call: extract factors from response ──
        record = _extract_record(response, model, prompt_hash, elapsed_ms)

        # ── Hand to witness (non-blocking) ──
        witness.record(record)

        # ── Return UNTOUCHED response to developer ──
        return response

    # Preserve the original method's signature for IDE support
    interceptor.__name__ = method_name
    interceptor.__qualname__ = f"{path}.{method_name}"
    interceptor.__doc__ = getattr(real_method, "__doc__", None)

    return interceptor


def _extract_prompt_text(messages: Any) -> str:
    """Extract hashable text from OpenAI message format.

    Handles:
        - List[dict] with "content" keys (standard chat format)
        - List[dict] with multimodal content (text parts only)
        - String input (simple prompt)
    """
    if isinstance(messages, str):
        return messages

    parts: list[str] = []
    if isinstance(messages, (list, tuple)):
        for msg in messages:
            if isinstance(msg, dict):
                content = msg.get("content", "")
                if isinstance(content, str):
                    parts.append(content)
                elif isinstance(content, list):
                    # Multimodal: extract text parts only
                    for part in content:
                        if isinstance(part, dict) and part.get("type") == "text":
                            parts.append(part.get("text", ""))
            elif hasattr(msg, "content"):
                # Pydantic model (openai types)
                c = getattr(msg, "content", "")
                if isinstance(c, str):
                    parts.append(c)

    return "\n".join(parts)


def _extract_record(
    response: Any,
    model: str,
    prompt_hash: str,
    elapsed_ms: int,
) -> InferenceRecord:
    """Extract an InferenceRecord from an OpenAI ChatCompletion response.

    Factor mapping:
        factor_a → SHA-256 of model identifier (model integrity)
        factor_b → latency in milliseconds (performance)
        factor_c → 0 if clean, 1 if refusal/violation detected
    """
    # Extract response text for hashing
    response_text = ""
    has_refusal = False

    # Handle ChatCompletion response object
    if hasattr(response, "choices") and response.choices:
        choice = response.choices[0]
        message = getattr(choice, "message", None)
        if message:
            content = getattr(message, "content", "") or ""
            response_text = content

            # Check for refusal (OpenAI's refusal field)
            refusal = getattr(message, "refusal", None)
            if refusal:
                has_refusal = True

        # Check finish_reason for content_filter
        finish_reason = getattr(choice, "finish_reason", "")
        if finish_reason == "content_filter":
            has_refusal = True

    # Extract token usage
    input_tokens = None
    output_tokens = None
    usage = getattr(response, "usage", None)
    if usage:
        input_tokens = getattr(usage, "prompt_tokens", None)
        output_tokens = getattr(usage, "completion_tokens", None)

    # Extract model info from response (may differ from request)
    actual_model = getattr(response, "model", model) or model
    system_fingerprint = getattr(response, "system_fingerprint", None)

    # Model hash: SHA-256 of model identifier + system_fingerprint (if available)
    model_hash_input = actual_model
    if system_fingerprint:
        model_hash_input = f"{actual_model}:{system_fingerprint}"
    model_hash = sha256_truncated(model_hash_input)

    # Response hash
    response_hash = sha256_truncated(response_text) if response_text else sha256_truncated("")

    return InferenceRecord(
        model_id=actual_model,
        model_hash=model_hash,
        prompt_hash=prompt_hash,
        response_hash=response_hash,
        latency_ms=elapsed_ms,
        input_tokens=input_tokens,
        output_tokens=output_tokens,
        has_refusal=has_refusal,
        provider="openai",
        system_fingerprint=system_fingerprint,
    )

File: packages/swt3-ai-ts/src/adapters/openai.ts
/**
 * SWT3 AI Witness SDK — OpenAI Adapter (ES6 Proxy).
 *
 * Uses JavaScript's native Proxy to intercept property access on the
 * OpenAI client, following the chain: client.chat.completions.create()
 *
 * Handles both:
 *   - Non-streaming: ChatCompletion response object
 *   - Streaming: AsyncIterable<ChatCompletionChunk> — accumulates chunks
 *     for hashing, then witnesses after stream completes
 *
 * The developer's code sees zero difference from using the raw client.
 */

import { sha256Truncated } from "../fingerprint.js";
import type { InferenceRecord } from "../types.js";
import type { Witness } from "../witness.js";

/**
 * Methods to intercept at each path level.
 * "chat.completions" → intercept "create"
 */
const INTERCEPT_PATHS: Record<string, Set<string>> = {
  "chat.completions": new Set(["create"]),
};

/**
 * Wrap an OpenAI client with an ES6 Proxy for transparent witnessing.
 */
export function wrapOpenAI(client: unknown, witness: Witness): unknown {
  return createProxy(client, witness, "");
}

function createProxy(target: unknown, witness: Witness, path: string): unknown {
  return new Proxy(target as object, {
    get(obj: object, prop: string | symbol): unknown {
      if (typeof prop === "symbol") return Reflect.get(obj, prop);

      const realValue = Reflect.get(obj, prop);
      const currentPath = path ? `${path}.${prop}` : prop;

      // Check if this property is an interceptable method
      if (path in INTERCEPT_PATHS && INTERCEPT_PATHS[path].has(prop)) {
        // This IS the method to intercept
        return createInterceptor(realValue as (...args: unknown[]) => unknown, witness);
      }

      // Check if this path could lead to interceptable methods
      const hasChildren = Object.keys(INTERCEPT_PATHS).some(
        (p) => p === currentPath || p.startsWith(currentPath + "."),
      );

      if (hasChildren && typeof realValue === "object" && realValue !== null) {
        return createProxy(realValue, witness, currentPath);
      }

      // Pass through
      return realValue;
    },
  });
}

function createInterceptor(
  realMethod: (...args: unknown[]) => unknown,
  witness: Witness,
): (...args: unknown[]) => unknown {
  return function interceptedCreate(this: unknown, ...args: unknown[]): unknown {
    const kwargs = (args[0] ?? {}) as Record<string, unknown>;
    const messages = kwargs.messages as unknown[];
    const model = (kwargs.model as string) ?? "unknown";
    const isStreaming = kwargs.stream === true;

    // Hash prompt before the call
    const promptText = extractPromptText(messages);
    const promptHash = sha256Truncated(promptText);

    // Hash system prompt separately (instruction drift detection)
    const systemPromptText = extractSystemPrompt(messages);
    const systemPromptHash = systemPromptText ? sha256Truncated(systemPromptText) : undefined;

    // Start latency timer
    const start = performance.now();

    // Call the real method (bound to its original this)
    const result = realMethod.call(this, ...args);

    if (isStreaming) {
      // Streaming: wrap the async iterable to accumulate chunks
      return handleStreaming(result, witness, model, promptHash, start, systemPromptHash);
    }

    // Non-streaming: result is a Promise<ChatCompletion>
    return (result as Promise<unknown>).then((response: unknown) => {
      const elapsedMs = Math.round(performance.now() - start);
      const record = extractRecord(response, model, promptHash, elapsedMs, systemPromptHash);
      witness.record(record);
      return response; // Return UNTOUCHED
    });
  };
}

// ── Streaming Handler ──────────────────────────────────────────────
//
// OpenAI's streaming returns an object with:
//   - Symbol.asyncIterator (for `for await...of`)
//   - .controller for abort
//   - .toReadableStream() for web streams
//
// We wrap the async iterator to accumulate content chunks, then
// witness after the stream completes. The developer's stream
// behavior is completely unchanged.

async function* streamAccumulator(
  stream: AsyncIterable<unknown>,
  witness: Witness,
  model: string,
  promptHash: string,
  startTime: number,
  systemPromptHash?: string,
): AsyncGenerator<unknown, void, undefined> {
  const textParts: string[] = [];
  let actualModel = model;
  let systemFingerprint: string | undefined;
  let hasRefusal = false;
  let finishReason = "";
  let inputTokens: number | undefined;
  let outputTokens: number | undefined;

  for await (const chunk of stream) {
    // Yield chunk to the developer immediately (untouched)
    yield chunk;

    // Extract data from chunk for witnessing
    const c = chunk as Record<string, unknown>;

    if (c.model) actualModel = c.model as string;
    if (c.system_fingerprint) systemFingerprint = c.system_fingerprint as string;

    const choices = c.choices as Array<Record<string, unknown>> | undefined;
    if (choices?.[0]) {
      const delta = choices[0].delta as Record<string, unknown> | undefined;
      if (delta?.content) {
        textParts.push(delta.content as string);
      }
      if (delta?.refusal) {
        hasRefusal = true;
      }
      if (choices[0].finish_reason) {
        finishReason = choices[0].finish_reason as string;
      }
    }

    // Usage info comes in the final chunk (stream_options: {include_usage: true})
    const usage = c.usage as Record<string, unknown> | undefined;
    if (usage) {
      inputTokens = usage.prompt_tokens as number | undefined;
      outputTokens = usage.completion_tokens as number | undefined;
    }
  }

  // Stream complete — witness the full inference
  const elapsedMs = Math.round(performance.now() - startTime);
  const responseText = textParts.join("");

  if (finishReason === "content_filter") hasRefusal = true;

  const modelHashInput = systemFingerprint
    ? `${actualModel}:${systemFingerprint}`
    : actualModel;

  const record: InferenceRecord = {
    modelId: actualModel,
    modelHash: sha256Truncated(modelHashInput),
    promptHash,
    responseHash: sha256Truncated(responseText),
    latencyMs: elapsedMs,
    inputTokens,
    outputTokens,
    guardrailsActive: 0,
    guardrailsRequired: 0,
    guardrailPassed: true,
    hasRefusal,
    provider: "openai",
    systemFingerprint,
    systemPromptHash,
    guardrailNames: [],
  };

  witness.record(record);
}

function handleStreaming(
  streamResult: unknown,
  witness: Witness,
  model: string,
  promptHash: string,
  startTime: number,
  systemPromptHash?: string,
): unknown {
  // The stream result is a Promise that resolves to the stream object
  // (OpenAI SDK returns Promise<Stream<ChatCompletionChunk>>)
  if (streamResult && typeof (streamResult as Promise<unknown>).then === "function") {
    return (streamResult as Promise<unknown>).then((stream: unknown) => {
      return wrapStream(stream, witness, model, promptHash, startTime, systemPromptHash);
    });
  }

  // Direct stream object (shouldn't happen but handle gracefully)
  return wrapStream(streamResult, witness, model, promptHash, startTime, systemPromptHash);
}

function wrapStream(
  stream: unknown,
  witness: Witness,
  model: string,
  promptHash: string,
  startTime: number,
  systemPromptHash?: string,
): unknown {
  const s = stream as Record<string | symbol, unknown>;

  // Create the accumulating async generator
  const gen = streamAccumulator(
    s as unknown as AsyncIterable<unknown>,
    witness,
    model,
    promptHash,
    startTime,
    systemPromptHash,
  );

  // Return a proxy that preserves all stream methods but overrides the iterator
  return new Proxy(s, {
    get(target: Record<string | symbol, unknown>, prop: string | symbol): unknown {
      // Override the async iterator
      if (prop === Symbol.asyncIterator) {
        return () => gen;
      }

      // Preserve other methods (controller, toReadableStream, etc.)
      const value = Reflect.get(target, prop);
      if (typeof value === "function") {
        return (value as Function).bind(target);
      }
      return value;
    },
  });
}

// ── Factor Extraction ──────────────────────────────────────────────

function extractSystemPrompt(messages: unknown): string | undefined {
  if (!Array.isArray(messages)) return undefined;

  const parts: string[] = [];
  for (const msg of messages) {
    if (typeof msg === "object" && msg !== null) {
      const m = msg as Record<string, unknown>;
      if (m.role !== "system") continue;
      const content = m.content;
      if (typeof content === "string") {
        parts.push(content);
      } else if (Array.isArray(content)) {
        for (const part of content) {
          if (typeof part === "object" && part !== null) {
            const p = part as Record<string, unknown>;
            if (p.type === "text" && typeof p.text === "string") {
              parts.push(p.text);
            }
          }
        }
      }
    }
  }
  return parts.length > 0 ? parts.join("\n") : undefined;
}

function extractPromptText(messages: unknown): string {
  if (typeof messages === "string") return messages;
  if (!Array.isArray(messages)) return "";

  const parts: string[] = [];
  for (const msg of messages) {
    if (typeof msg === "object" && msg !== null) {
      const m = msg as Record<string, unknown>;
      const content = m.content;
      if (typeof content === "string") {
        parts.push(content);
      } else if (Array.isArray(content)) {
        for (const part of content) {
          if (typeof part === "object" && part !== null) {
            const p = part as Record<string, unknown>;
            if (p.type === "text" && typeof p.text === "string") {
              parts.push(p.text);
            }
          }
        }
      }
    }
  }
  return parts.join("\n");
}

function extractRecord(
  response: unknown,
  model: string,
  promptHash: string,
  elapsedMs: number,
  systemPromptHash?: string,
): InferenceRecord {
  const r = response as Record<string, unknown>;
  let responseText = "";
  let hasRefusal = false;

  const choices = r.choices as Array<Record<string, unknown>> | undefined;
  if (choices?.[0]) {
    const message = choices[0].message as Record<string, unknown> | undefined;
    if (message) {
      responseText = (message.content as string) ?? "";
      if (message.refusal) hasRefusal = true;
    }
    if (choices[0].finish_reason === "content_filter") hasRefusal = true;
  }

  let inputTokens: number | undefined;
  let outputTokens: number | undefined;
  const usage = r.usage as Record<string, unknown> | undefined;
  if (usage) {
    inputTokens = usage.prompt_tokens as number | undefined;
    outputTokens = usage.completion_tokens as number | undefined;
  }

  const actualModel = (r.model as string) ?? model;
  const systemFingerprint = r.system_fingerprint as string | undefined;

  const modelHashInput = systemFingerprint
    ? `${actualModel}:${systemFingerprint}`
    : actualModel;

  return {
    modelId: actualModel,
    modelHash: sha256Truncated(modelHashInput),
    promptHash,
    responseHash: sha256Truncated(responseText),
    latencyMs: elapsedMs,
    inputTokens,
    outputTokens,
    guardrailsActive: 0,
    guardrailsRequired: 0,
    guardrailPassed: true,
    hasRefusal,
    provider: "openai",
    systemFingerprint,
    systemPromptHash,
    guardrailNames: [],
  };
}
